rgw: delete zone/pools for multisite configuation (#10728)
rook/rook: design/ceph/object/zone.md

When the storage admin is ready to sync data from another Ceph cluster with multisite set up (primary cluster) to a Rook Ceph cluster (pulling cluster), the pulling cluster will have a newly created in the zone group from the primary cluster.

A [ceph-object-pull-realm](/design/ceph/object/ceph-object-pull-realm.md) resource must be created to pull the realm information from the primary cluster to the pulling cluster.

Once the ceph-object-pull-realm is configured a ceph-object-zone must be created.

After an ceph-object-store is configured to be in this ceph-object-zone, the all Ceph multisite resources will be running and data between the two clusters will start syncing.

## Deleting and Reconfiguring the Ceph Object Zone

At the moment creating a CephObjectZone resource does not handle configuration updates for the zone.

By default when a CephObjectZone is deleted, the pools supporting the zone are not deleted from the Ceph cluster.

> If `preservePoolsOnDelete` is set to false, then the CephObjectZone will be removed only if all CephObjectStores that reference this zone are deleted first.

Reviewer: This is not the behavior we agreed upon in our last meeting. Pool deletion should be independent of zone deletion.

```suggestion
If `preservePoolsOnDelete` is set to false, the pools supporting the zone are deleted from the Ceph cluster. <!-- and this should be paired with the above statement -->

A CephObjectZone will be removed only if all CephObjectStores that reference the zone are deleted first.
```

At the moment creating an ceph-object-zone resource only handles Day 1 initial configuration for the zone.

> By default when the ceph-object-zone resource is deleted or modified, the zone or pools is not deleted from the Ceph cluster.

Reviewer: ```suggestion
By default when a CephObjectZone is deleted or modified, the poos supporting the zone are not deleted from the Ceph cluster.
```

If user disables `preservePoolsOnDelete` then zone or pool will be removed only if following conditions satisfied:
- all the objects belong to this zone need to deleted first.
- it should not be master zone.

> - if it is master zone, then no other zones belongs to zonegroup.

Reviewer: Does this mean deletion behavior of the zone itself is different depending on the `preservePoolsOnDelete` setting? That doesn't seem quite right. That setting should just affect pools (and be mentioned in the paragraph above). The behavior of deleting the zone itself can be talked about here.

> 

Author: @cbodley: Is okay to remove pools of zone if above conditions satisy? This is mostly executed in case of clean-up. All the data residing on the pool will be removed when `objectstores` are removed. Do I need to add more conditions here?

Reviewer 2: these look like a good set of conditions, but it's not clear to me where these are being enforced

if there are still objectstores referring to the zone resource, we should prevent the zone resource from deleted in the first place

if we're trying to delete a zone resource and find that there are other zones in its period configuration, we should remove that zone from its zonegroup and commit the period

if all of the above is enforced, then i think it's fine to delete pools during zone resource deletion

Author: We are checking dependency before deleting the zone in [CephObjectZoneDependents](https://github.com/rook/rook/pull/10728/files#diff-2acd79f015dc6bbe01d58404a344555269d205e7ca287996bd11595794dd196cR35) and zone deletion will be blocked until dependency are removed

At the moment creating an ceph-object-zone resource only handles Day 1 initial configuration for the zone.

By default when a CephObjectZone is deleted or modified, the pools supporting the zone are not deleted from the Ceph cluster.

If user disables `preservePoolsOnDelete` then zone or pool will be removed only if all the objectstores belong to this zone need to deleted first.

### Deleting ObjectStore belong to a multisite

If the `objectstore` belong to [master zone](/design/ceph/object/ceph-multisite-overview.md/#master-zone) then deletion will be blocked until following is satisfied:
- either all the peer zone need to removed first or choose another zone as master as steps mentioned [here](https://docs.ceph.com/en/latest/radosgw/multisite/#failover-and-disaster-recovery)
- if no more peers attached then user need to remove the residing data.

> If the object store belong to peer zone, then it is directly removed, assuming data is synced and safely hold in other peers.

Reviewer 3: ```suggestion
If the object store belongs to a peer zone, then the deletion is not blocked and the zone is immediately deleted. If the data is still required, it will still be available in the peer cluster.
```

- `customEndpoints`:  Specify the endpoint(s) that will accept multisite replication traffic for this zone. You may include the port in the definition if necessary. For example: "https://my-object-store.my-domain.net:443".

> - `preservePoolsOnDelete`: If it is set to 'true' the pools used to support the zone will remain when the ceph object zone CR will be deleted. This is a security measure to avoid accidental loss of data. It is set to 'true' by default. If not specified is also deemed as 'true'.

Reviewer 3: Should the default be `false` if it's not a master zone? It's not critical to preserve the pools in that case, right?

Author: @travisn: That's true. Are suggesting the following:
If user sets the value then directly use that value, if not have a helper function to determine the value depending on the master or peer zone than using the inbuild Kubernetes default value.
One another reason to set default value as `true` to preseve the old behaviour of pools not getting removed from the backend even if Zone CR is removed. @alimaredia please share your opinion

- `customEndpoints`:  Specify the endpoint(s) that will accept multisite replication traffic for this zone. You may include the port in the definition if necessary. For example: "https://my-object-store.my-domain.net:443".

> - `preservePoolsOnDelete`: If it is set to 'true' the pools used to support the zone will remain when the ceph object zone CR will be deleted. This is a security measure to avoid accidental loss of data. It is set to 'true' by default. If not specified is also deemed as 'true'.

Reviewer: ```suggestion
- `preservePoolsOnDelete`: If it is set to 'true' the pools used to support the zone will remain when the CephObjectZone is deleted. This is a security measure to avoid accidental loss of data. It is set to 'true' by default. If not specified it is also deemed as 'true'.
```
